### PR TITLE
The "Log editor data" button should copy to the clipboard if clicked with Alt/Option

### LINKED ISSUE #136: The button that logs editor data to console should allow copying it to clipboard
If you want to quickly use the editor data, e.g. put it in `editor.setData()` again or paste it in a GH issue (which happens a lot), there's one unnecessary step. 

Also, when the content is huge (GHS, something from Word), extracting stuff from the logged text in the console is very inconvenient.

My idea is that the button should change its functionality using a modifier key.

### PR DESCRIPTION
**Pull request by oleq**
Feature: The "Log editor data" button should copy to the clipboard if clicked with Alt/Option. Closes #136.

### CODE REVIEW

File: src/editorquickactions.css [hunk 1]
```diff
@@ -13,5 +13,37 @@
 
 	& > .ck-inspector-button  {
 		margin-left: .3em;
+
+		/* The tick icon shows up when the content was copied to clipboard */
```
Comment (przemyslaw-zan) [inline src/editorquickactions.css:17]:
```suggestion
		/* The checkmark icon shows up when the content was copied to clipboard */
```

File: src/editorquickactions.js [hunk 2]
```diff
@@ -3,33 +3,46 @@
...
+import copy from 'copy-to-clipboard';
 
 import Button from './components/button';
 
 import SourceIcon from './assets/img/source.svg';
 import ConsoleIcon from './assets/img/console.svg';
 import ReadOnlyIcon from './assets/img/read-only.svg';
 import TrashIcon from './assets/img/trash.svg';
+import ClipboardIcon from './assets/img/clipboard.svg';
+import TickIcon from './assets/img/tick.svg';
```
Comment (przemyslaw-zan) [inline src/editorquickactions.js:19]:
```suggestion
import CopyToClipboardIcon from './assets/img/copy-to-clipboard.svg';
import CheckmarkIcon from './assets/img/checkmark.svg';
```

File: src/editorquickactions.js [hunk 3]
```diff
@@ -45,6 +58,72 @@ class EditorQuickActions extends Component {
...
+	}
+
+	_getLogButton() {
+		let icon, text;
+
+		if ( this.state.wasEditorDataJustCopied ) {
+			icon = <TickIcon />;
+			text = 'Data copied to clipboard.';
+		} else {
+			icon = this.state.isAltKeyPressed ? <ClipboardIcon /> : <SourceIcon />;
```
Comment (przemyslaw-zan) [inline src/editorquickactions.js:83]:
```suggestion
			icon = this.state.isAltKeyPressed ? <CopyToClipboardIcon /> : <SourceIcon />;
```

File: tests/inspector/editorquickactions.js [hunk 4]
```diff
@@ -12,6 +12,10 @@ import { Provider } from 'react-redux';
 
 import EditorQuickActions from '../../src/editorquickactions';
 
+import SourceIcon from '../../src/assets/img/source.svg';
+import ClipboardIcon from '../../src/assets/img/clipboard.svg';
+import TickIcon from '../../src/assets/img/tick.svg';
```
Comment (przemyslaw-zan) [inline tests/inspector/editorquickactions.js:17]:
```suggestion
import CopyToClipboardIcon from '../../src/assets/img/copy-to-clipboard.svg';
import CheckmarkIcon from '../../src/assets/img/checkmark.svg';
```

File: tests/inspector/editorquickactions.js [hunk 5]
```diff
@@ -87,6 +94,171 @@ describe( '<EditorQuickActions />', () => {
...
+				document.dispatchEvent( new KeyboardEvent( 'keydown', { key: 'Alt' } ) );
+				wrapper.update();
+
+				expect( quickActions.state() ).to.deep.equal( {
+					isAltKeyPressed: true,
+					wasEditorDataJustCopied: false
+				} );
+
+				logButton = wrapper.find( 'Button' ).at( 1 );
+				expect( logButton.props().icon.type ).to.equal( ClipboardIcon );
```
Comment (przemyslaw-zan) [inline tests/inspector/editorquickactions.js:119]:
```suggestion
				expect( logButton.props().icon.type ).to.equal( CopyToClipboardIcon );
```

File: tests/inspector/editorquickactions.js [hunk 6]
```diff
@@ -87,6 +94,171 @@ describe( '<EditorQuickActions />', () => {
...
+
+				wrapper.update();
+
+				expect( quickActions.state() ).to.deep.equal( {
+					isAltKeyPressed: true,
+					wasEditorDataJustCopied: true
+				} );
+
+				logButton = wrapper.find( 'Button' ).at( 1 );
+				expect( logButton.props().icon.type ).to.equal( TickIcon );
```
Comment (przemyslaw-zan) [inline tests/inspector/editorquickactions.js:151]:
```suggestion
				expect( logButton.props().icon.type ).to.equal( CheckmarkIcon );
```

File: tests/inspector/editorquickactions.js [hunk 7]
```diff
@@ -87,6 +94,171 @@ describe( '<EditorQuickActions />', () => {
...
+				expect( logButton.props().text ).to.equal( 'Data copied to clipboard.' );
+
+				// Make sure the tick icon + text stay for 3000ms.
```
Comment (przemyslaw-zan) [inline tests/inspector/editorquickactions.js:154]:
```suggestion
				// Make sure the checkmark icon + text stay for 3000ms.
```

File: tests/inspector/editorquickactions.js [hunk 8]
```diff
@@ -87,6 +94,171 @@ describe( '<EditorQuickActions />', () => {
...
+				clock.tick( 2500 );
+				wrapper.update();
+
+				expect( quickActions.state() ).to.deep.equal( {
+					isAltKeyPressed: true,
+					wasEditorDataJustCopied: true
+				} );
+
+				logButton = wrapper.find( 'Button' ).at( 1 );
+				expect( logButton.props().icon.type ).to.equal( TickIcon );
```
Comment (przemyslaw-zan) [inline tests/inspector/editorquickactions.js:164]:
```suggestion
				expect( logButton.props().icon.type ).to.equal( CheckmarkIcon );
```

File: tests/inspector/editorquickactions.js [hunk 9]
```diff
@@ -87,6 +94,171 @@ describe( '<EditorQuickActions />', () => {
...
+				expect( logButton.props().text ).to.equal( 'Data copied to clipboard.' );
+
+				// Wait for the tick icon + text to disappear.
```
Comment (przemyslaw-zan) [inline tests/inspector/editorquickactions.js:167]:
```suggestion
				// Wait for the checkmark icon + text to disappear.
```

File: tests/inspector/editorquickactions.js [hunk 10]
```diff
@@ -87,6 +94,171 @@ describe( '<EditorQuickActions />', () => {
...
+				clock.tick( 1000 );
+				wrapper.update();
+
+				expect( quickActions.state() ).to.deep.equal( {
+					isAltKeyPressed: true,
+					wasEditorDataJustCopied: false
+				} );
+
+				logButton = wrapper.find( 'Button' ).at( 1 );
+				expect( logButton.props().icon.type ).to.equal( ClipboardIcon );
```
Comment (przemyslaw-zan) [inline tests/inspector/editorquickactions.js:177]:
```suggestion
				expect( logButton.props().icon.type ).to.equal( CopyToClipboardIcon );
```

**Review by przemyslaw-zan — CHANGES_REQUESTED** [review]
Perhaps `svg` names could be clearer?

- `clipboard` -> `copy-to-clipboard`
- `tick` -> `checkmark`

As far as functionality is concerned, it seems to works as expected. However, if `alt` is pressed and then released without using the button, Chrome switches focus to options button, and Firefox toggles options bar. I suppose we don't want to intercept that, since that is default key behavior on these browsers, but perhaps we could consider different key than `alt`? `shift` and `ctrl` do not trigger any additional behavior. Not sure about that one, maybe i'm looking too deep into this.

**Comment by oleq** [comment]
Thanks for the review @przemyslaw-zan.

> As far as functionality is concerned, it seems to works as expected. However, if `alt` is pressed and then released without using the button, Chrome switches focus to options button, and Firefox toggles options bar, both of which is somewhat annoying - this could cause poor user experience IMO.

What do you mean by "options button" and "options bar"? Is it the UI of the web browser? Because I only tested on Chrome/FF on Mac and there's nothing exceptional happening. Is it an OS thing maybe?

**Comment by oleq** [comment]
As for the changes in the code, feel free to commit them straight to the PR. They make sense :)

**Comment by przemyslaw-zan** [comment]
@oleq sorry for not specifying, yes, i meant UI of the browsers, and the behavior happens on Windows

**Comment by przemyslaw-zan** [comment]
Firefox: (Options bar toggles on and off)



Chrome: (Options button takes focus and prevents further use)

**Comment by oleq** [comment]
I didn't anticipate that. Nice catch 👍 

I see three options now:

*   Stick with Alt (because it's alternate after all) but `preventDefault` the event
    *   👎 This will break native browser features that you mentioned,
    *   👎 This would also break other features in the editor like typing accents (ąęśćżźćół...),
        *   OTOH... I think I already detect when the Alt key is being pressed "solo" only, so maybe this would not happen. Needs to be checked.
*   Stick with Alt but `preventDefault` the event only when the inspector is being hovered
    *   👍 Not breaking native browser features most of the time,
    *   👍 Not breaking accents (most of the time)
        *   Or not at all, see my comment in the previous strategy.
    *   👎 Yet breaking both 👆 occasionally (accidentally), which is not a great UX. Quite a surprise when sometimes you can type things and sometimes you cannot.
    *   👎 Annoying if you move the cursor from the top of the page into inspector because the button will change the icon very late.
    *   👎 More complexity in the code.
*   Switch to Shift or Ctrl
    *   👍 Not breaking anything this time, I hope.
    *   👍 Low cost of implementation.
    *   👎 Not very intuitive because clearly, the alternate key should... alternate the features. It's a very common pattern, at least on Mac.

Which sounds best to you? I'm lazy so I'd rather go with Shift/Ctrl if confirmed safe.

**Comment by przemyslaw-zan** [comment]
I agree with you on the last option being the best. Personally, I think that `ctrl` is the next best choice.

**Comment by oleq** [comment]
FYI: I used `Shift` instead of `Ctrl`. Turns out, Ctrl+Click is already intercepted and opens the native "right click" context menu. If I went with Ctrl, I'd have to `preventDefault` the `'contextmenu'` event which sounds like unnecessary trouble.

**Review by przemyslaw-zan — APPROVED** [review]
LGTM

**Pull request merged**